File: backend/lyrics/parser.js
import * as cheerio from 'cheerio';

/**
 * Extracts plain text lyrics from a Genius song page HTML.
 * @param {string} html - The HTML response from Genius.com.
 * @returns {string} The cleaned plain text lyrics.
 */
export function extractLyrics(html) {
  const $ = cheerio.load(html);
  let lyrics = '';

  // 1. Modern Genius layout: elements with data-lyrics-container="true"
  const containers = $('div[data-lyrics-container="true"]');
  if (containers.length > 0) {
    containers.each((_, container) => {
      const elem = $(container);
      
      // Replace <br> tags with newlines before extracting text
      elem.find('br').replaceWith('\n');
      
      // Inject spacing for paragraphs or nested divs to preserve line breaks
      elem.find('p, div').prepend('\n').append('\n');
      
      lyrics += elem.text() + '\n';
    });
  }

  // 2. Legacy Genius layout: element with class "lyrics"
  if (!lyrics.trim()) {
    const legacyContainer = $('.lyrics');
    if (legacyContainer.length > 0) {
      legacyContainer.find('br').replaceWith('\n');
      lyrics = legacyContainer.text();
    }
  }

  // 3. Alternate/fallback Genius layout: classes prefixed with Lyrics__Container
  if (!lyrics.trim()) {
    $('[class*="Lyrics__Container"]').each((_, container) => {
      const elem = $(container);
      elem.find('br').replaceWith('\n');
      lyrics += elem.text() + '\n';
    });
  }

  return cleanLyrics(lyrics);
}

/**
 * Cleans extracted lyric text (removing multiple newlines, trimming lines).
 * @param {string} lyrics - Raw lyric text.
 * @returns {string} Cleaned lyric text.
 */
function cleanLyrics(lyrics) {
  if (!lyrics) return '';
  return lyrics
    .split('\n')
    .map(line => line.trim())
    // Remove lyrics page headers like "Lyrics", "About", "1 Contributor"
    .filter(line => !/^\d+\s+Contributor(s)?$/i.test(line))
    .filter(line => !/^You\s+may\s+also\s+like$/i.test(line))
    .join('\n')
    .replace(/\n{3,}/g, '\n\n') // Collapse excessive newlines down to max 2
    .trim();
}

File: backend/lyrics/parser.py
import re
from bs4 import BeautifulSoup

def extract_lyrics(html: str) -> str:
    """
    Scrapes and extracts plain text lyrics from Genius song page HTML.
    """
    soup = BeautifulSoup(html, 'html.parser')
    lyrics = ""

    # 1. Modern Genius layout: div[data-lyrics-container="true"]
    containers = soup.find_all('div', attrs={"data-lyrics-container": "true"})
    if containers:
        for container in containers:
            # Replace <br> with newlines to preserve lines
            for br in container.find_all('br'):
                br.replace_with('\n')
            # Extract plain text
            lyrics += container.get_text() + '\n'

    # 2. Legacy Genius layout: div.lyrics
    if not lyrics.strip():
        legacy = soup.find('div', class_='lyrics')
        if legacy:
            for br in legacy.find_all('br'):
                br.replace_with('\n')
            lyrics = legacy.get_text()

    # 3. Alternate class contains "Lyrics__Container"
    if not lyrics.strip():
        for container in soup.find_all(class_=re.compile(r'Lyrics__Container')):
            for br in container.find_all('br'):
                br.replace_with('\n')
            lyrics += container.get_text() + '\n'

    return clean_lyrics(lyrics)

def clean_lyrics(lyrics: str) -> str:
    """
    Trims, cleans up metadata lines, and formats newlines in the raw lyrics.
    """
    if not lyrics:
        return ""
    
    lines = lyrics.split('\n')
    cleaned_lines = []
    
    for line in lines:
        line_stripped = line.strip()
        # Filter metadata noise
        if re.match(r'^\d+\s+Contributor(s)?$', line_stripped, re.IGNORECASE):
            continue
        if re.match(r'^You\s+may\s+also\s+like$', line_stripped, re.IGNORECASE):
            continue
        cleaned_lines.append(line_stripped)
        
    cleaned_text = '\n'.join(cleaned_lines)
    # Collapse multiple consecutive newlines down to max 2
    cleaned_text = re.sub(r'\n{3,}', '\n\n', cleaned_text)
    return cleaned_text.strip()
